// @sentry/nuxt@8.26.0/build/cjs/module.js
const fs = require('fs');
const path = require('path');
const kit = require('@nuxt/kit');
const sourceMaps = require('./vite/sourceMaps.js');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
const module$1 = kit.defineNuxtModule({
  meta: {
    name: '@sentry/nuxt/module',
    configKey: 'sentry',
    compatibility: {
      nuxt: '^3.0.0',
    },
  },
  defaults: {},
  setup(moduleOptions, nuxt) {
    const moduleDirResolver = kit.createResolver((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('module.js', document.baseURI).href)));
    const buildDirResolver = kit.createResolver(nuxt.options.buildDir);

    const clientConfigFile = findDefaultSdkInitFile('client');

    if (clientConfigFile) {
      // Inject the client-side Sentry config file with a side effect import
      kit.addPluginTemplate({
        mode: 'client',
        filename: 'sentry-client-config.mjs',
        getContents: () =>
          `import "${buildDirResolver.resolve(`/${clientConfigFile}`)}"\n` +
          'import { defineNuxtPlugin } from "#imports"\n' +
          'export default defineNuxtPlugin(() => {})',
      });

      kit.addPlugin({ src: moduleDirResolver.resolve('./runtime/plugins/sentry.client'), mode: 'client' });
    }

    const serverConfigFile = findDefaultSdkInitFile('server');

    if (serverConfigFile) {
      // Inject the server-side Sentry config file with a side effect import
      kit.addPluginTemplate({
        mode: 'server',
        filename: 'sentry-server-config.mjs',
        getContents: () =>
          `import "${buildDirResolver.resolve(`/${serverConfigFile}`)}"\n` +
          'import { defineNuxtPlugin } from "#imports"\n' +
          'export default defineNuxtPlugin(() => {})',
      });

      kit.addServerPlugin(moduleDirResolver.resolve('./runtime/plugins/sentry.server'));
    }

    if (clientConfigFile || serverConfigFile) {
      sourceMaps.setupSourceMaps(moduleOptions, nuxt);
    }
  },
});

function findDefaultSdkInitFile(type) {
  const possibleFileExtensions = ['ts', 'js', 'mjs', 'cjs', 'mts', 'cts'];

  const cwd = process.cwd();
  const filePath = possibleFileExtensions
    .map(e =>
      path.resolve(
        type === 'server'
          ? path.join(cwd, 'public', `instrument.${type}.${e}`)
          : path.join(cwd, `sentry.${type}.config.${e}`),
      ),
    )
    .find(filename => fs.existsSync(filename));

  return filePath ? path.basename(filePath) : undefined;
}

module.exports = module$1;
