// @sentry/nuxt@8.25.0/build/esm/module.js
import * as fs from 'fs';
import * as path from 'path';
import { defineNuxtModule, createResolver, addPluginTemplate, addPlugin, addServerPlugin } from '@nuxt/kit';
import { setupSourceMaps } from './vite/sourceMaps.js';

const module = defineNuxtModule({
  meta: {
    name: '@sentry/nuxt/module',
    configKey: 'sentry',
    compatibility: {
      nuxt: '^3.0.0',
    },
  },
  defaults: {},
  setup(moduleOptions, nuxt) {
    const moduleDirResolver = createResolver(import.meta.url);
    const buildDirResolver = createResolver(nuxt.options.buildDir);

    const clientConfigFile = findDefaultSdkInitFile('client');

    if (clientConfigFile) {
      // Inject the client-side Sentry config file with a side effect import
      addPluginTemplate({
        mode: 'client',
        filename: 'sentry-client-config.mjs',
        getContents: () =>
          `import "${buildDirResolver.resolve(`/${clientConfigFile}`)}"\n` +
          'export default defineNuxtPlugin(() => {})',
      });

      addPlugin({ src: moduleDirResolver.resolve('./runtime/plugins/sentry.client'), mode: 'client' });
    }

    const serverConfigFile = findDefaultSdkInitFile('server');

    if (serverConfigFile) {
      // Inject the server-side Sentry config file with a side effect import
      addPluginTemplate({
        mode: 'server',
        filename: 'sentry-server-config.mjs',
        getContents: () =>
          `import "${buildDirResolver.resolve(`/${serverConfigFile}`)}"\n` +
          'export default defineNuxtPlugin(() => {})',
      });

      addServerPlugin(moduleDirResolver.resolve('./runtime/plugins/sentry.server'));
    }

    if (clientConfigFile || serverConfigFile) {
      setupSourceMaps(moduleOptions, nuxt);
    }
  },
});

function findDefaultSdkInitFile(type) {
  const possibleFileExtensions = ['ts', 'js', 'mjs', 'cjs', 'mts', 'cts'];

  const cwd = process.cwd();
  const filePath = possibleFileExtensions
    .map(e =>
      path.resolve(
        type === 'server'
          ? path.join(cwd, 'public', `instrument.${type}.${e}`)
          : path.join(cwd, `sentry.${type}.config.${e}`),
      ),
    )
    .find(filename => fs.existsSync(filename));

  return filePath ? path.basename(filePath) : undefined;
}

export { module as default };
